### PR TITLE
Update readme

### PR DESCRIPTION
**Pull request by christianbean**

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -158,5 +176,217 @@ using the ``gridded_perms_of_length`` method.
...
+--------------------------------
+
+TileScope can be imported in a interactive Python session from ``tilings.tilescope``.
+
+.. code:: python
+
+       >>> from tilings.tilescope import *
+
+Importing ``*`` from ``tilings.tilescope`` supplies you with the "TileScope" and
+‘TileScopePack’ classes. Running the TileScope is as simple as choosing a class
```
Comment (enadeau) [inline README.rst:235]: The bracket are inconsistent

File: README.rst [hunk 2]
```diff
@@ -158,5 +176,217 @@ using the ``gridded_perms_of_length`` method.
...
+
+.. code:: python
+
+       >>> [spec.count_objects_of_size(i) for i in range(10)]
+       [1, 1, 2, 5, 14, 42, 132, 429, 1430, 4862]
+
+Of course we see the Catalan numbers! We can also sample uniformly using the
+``random_sample_object_of_size`` method. This will return a ``GriddedPerm``. If
+you want the underlying ``Perm``, this can be accessed with ``patt`` attribute.
+I have done this here, and then used the ``permuta.Perm.ascii_plot`` method for
```
Comment (enadeau) [inline README.rst:329]: We should probably avoid using first person

File: README.rst [hunk 3]
```diff
@@ -158,5 +176,217 @@ using the ``gridded_perms_of_length`` method.
...
+us to visualise it.
+
+.. code:: python
+
+       >>> gp = spec.random_sample_object_of_size(10)
+       >>> perm = gp.patt
+       >>> print(perm)
+       5042136987
+       >>> print(perm.ascii_plot())
+       | | | | | | | | | |
```
Comment (enadeau) [inline README.rst:339]: The line of the grid are not aligned properly. Is this a bug on permuta?

File: README.rst (hunk 2 above)
Comment (christianbean) [reply README.rst:329]: are we ok with 'we'?

File: README.rst (hunk 3 above)
Comment (christianbean) [reply README.rst:339]: No, just a copy-paste error

File: README.rst (hunk 2 above)
Comment (enadeau) [reply README.rst:329]: Yes it's how you wrote most of the README so the first person sounded weird

File: README.rst [hunk 4]
```diff
@@ -158,5 +179,900 @@ using the ``gridded_perms_of_length`` method.
...
+       1: Av(120)             1: Av+(120)
+                              Requirement 0:
+                              0: (0, 0)
+       -------
+       1 -> ()
+       is atom
+       +-+
+       | |
+       +-+
+       <BLANKLINE>
```
[the remark this answers is not in the document]
Comment (jaypantone) [reply README.rst:283]: As everyone knows, I feel very strongly against changing code just to make the doctest happy for moral reasons! I think it's fine to leave it there.

[the remark this answers is not in the document]
Comment (christianbean) [reply README.rst:283]: I tend to agree. If you had seen what it was like before you started your review, you would have been even more mad! The cool head of Emile got me through it.

File: README.rst [hunk 5]
```diff
@@ -31,25 +41,26 @@ To install ``tilings`` on your system, run:
 
        pip install tilings
 
-It is also possible to install ``tilings`` in development mode to work
-on the source code, in which case you run the following after cloning
-the repository:
+If you would like tot edit the source code, you should install ``tilings`` in
```
Comment (ulfarsson) [inline README.rst:44]: tot -> to

File: README.rst [hunk 6]
```diff
@@ -158,5 +179,900 @@ using the ``gridded_perms_of_length`` method.
...
+system of equations implied by the specification.
+
+.. code:: python
+
+       >>> list(spec.get_equations())
+       [Eq(F_0(x), F_1(x) + F_2(x)), Eq(F_1(x), 1), Eq(F_2(x), F_3(x)), Eq(F_3(x), F_0(x)**2*F_4(x)), Eq(F_4(x), x)]
+
+You can also pass these directly to the ``solve`` method in ``sympy`` by using the
+``get_genf`` method. It will then return the solution which matches the initial
+conditions.
```
[the remark this answers is not in the document]
Comment (christianbean) [reply README.rst:377]: I have added a few sentences and a easy to use get_maple_equations method. Better?

File: README.rst [hunk 7]
```diff
@@ -158,5 +179,900 @@ using the ``gridded_perms_of_length`` method.
...
+
+       >>> spec.get_genf()
+       (1 - sqrt(1 - 4*x))/(2*x)
+
+
+StrategyPacks
+=============
+
+Each strategy pack is essentially a different algorithm. These can be accessed
+as class methods on ``TileScopePack``. They are
```
[the remark this answers is not in the document]
Comment (christianbean) [reply README.rst:389]: Added a few new lines, do you agree?

File: README.rst [hunk 8]
```diff
@@ -158,5 +179,900 @@ using the ``gridded_perms_of_length`` method.
...
+       ●: point                                 ●: point
+       Crossing obstructions:                   Requirement 0:
+       0123: (0, 0), (0, 0), (0, 0), (0, 1)     0: (0, 1)
+       0132: (0, 0), (0, 0), (0, 1), (0, 0)
+       0312: (0, 0), (0, 1), (0, 0), (0, 0)
+       3012: (0, 1), (0, 0), (0, 0), (0, 0)
+       Requirement 0:
+       0: (0, 1)
+
+  # TODO: explain the simplification
```
[the remark this answers is not in the document]
Comment (christianbean) [reply README.rst:986]: Added a short paragraph. Do you think an example would be clearer? I could add a code snippet using the add_obstruction method?

File: README.rst [hunk 9]
```diff
@@ -158,5 +179,900 @@ using the ``gridded_perms_of_length`` method.
...
+       | |●| |            \: Av(01)     ●: point           /: Av(10)
+       +-+-+-+                          Requirement 0:
+       /: Av(10)                        0: (0, 0)
+       \: Av(01)
+       ●: point
+       Requirement 0:
+       0: (1, 0)
+
+We instead use the symbol ``*`` to make us aware that this is not a Cartesian
+product, but we must also count the possible interleavings.
```
[the remark this answers is not in the document]
Comment (christianbean) [reply README.rst:950]: I changed it to match the above blurb

 "to make us aware that this is not counted
by the Cauchy product"

File: README.rst [hunk 10]
```diff
@@ -158,5 +179,900 @@ using the ``gridded_perms_of_length`` method.
...
+       1: Av(120)
+       ●: point
+       Requirement 0:
+       0: (1, 2)
+
+The ``x`` in the printed above rule is used to denote Cartesian product. To
+guarantee that these rules are always a Cartesian product we must also ensure
+any two cells on the same row or column are also contained in the same factor,
+otherwise when counting the left hand side we have to consider the possible
+interleavings going on.
```
[the remark this answers is not in the document]
Comment (christianbean) [reply README.rst:908]: I rephrased this paragraph, but the section as a whole according to your comment. Could you take another look?

File: README.rst [hunk 11]
```diff
@@ -158,5 +179,900 @@ using the ``gridded_perms_of_length`` method.
...
+       >>> for rule in strategy_generator(tiling):
+       ...     print(rule)
+       fuse columns 0 and 1
+       +-+-+                      +-+
+       |\|\|                   ↣  |\|
+       +-+-+                      +-+
+       \: Av(01)                  \: Av(01)
+       Crossing obstructions:     Assumption 0:
+       01: (0, 0), (1, 0)         can count occurrences of
+                                  0: (0, 0)
```
[the remark this answers is not in the document]
Comment (christianbean) [reply README.rst:1036]: Added.

File: README.rst (hunk 5 above)
Comment (christianbean) [reply README.rst:44]: fixed

File: README.rst [hunk 12]
```diff
@@ -128,25 +142,32 @@ have a tiling that corresponds to non-empty permutation avoiding
        012: (0, 0), (2, 0), (2, 0)
        Requirement 0:
        0: (1, 1)
-       >>> til.dimensions
+       >>> tiling.dimensions
```
Comment (ulfarsson) [inline README.rst:145]: Maybe stop this code cell before calling the dimensions. And add text before continuing another code cell, e.g., "There are several properties of tilings you can call... e.g.,"

File: README.rst [hunk 13]
```diff
@@ -128,25 +142,32 @@ have a tiling that corresponds to non-empty permutation avoiding
...
-       >>> sorted(til.possibly_empty)
+       >>> sorted(tiling.possibly_empty)
        [(0, 0), (2, 0)]
-       >>> til.positive_cells
+       >>> tiling.positive_cells
        frozenset({(1, 1)})
 
+Those who have read ahead, or already started using tilings may have noticed
+that a ``Tiling`` can also be defined with a third argument called ``assumptions``.
+These can be used to keep track of occurrences gridded permutations on
```
Comment (ulfarsson) [inline README.rst:158]: OF gridded perms

File: README.rst [hunk 14]
```diff
@@ -158,5 +179,938 @@ using the ``gridded_perms_of_length`` method.
        102: (0, 0), (0, 0), (1, 1)
 
 There are numerous other methods and properties. Many of these specific
```
Comment (ulfarsson) [inline README.rst:181]: ARE specific

File: README.rst [hunk 15]
```diff
@@ -158,5 +179,938 @@ using the ``gridded_perms_of_length`` method.
...
+       tilescope spec -h
+
+To search for a combinatorial specification use the subcommand
+``tilescope spec``, e.g.
+
+.. code:: bash
+
+       tilescope spec 231 point_placements
+
+It will always try to solve for the generating function, although in some
```
Comment (ulfarsson) [inline README.rst:221]: By default the algorithm will ...

File: README.rst [hunk 16]
```diff
@@ -158,5 +179,938 @@ using the ``gridded_perms_of_length`` method.
...
+cases you will come across some not-yet-implemented features; for more information
+please join us on our `Discord server <https://discord.gg/ySJD6SV>`__,
+where we'd be happy to talk about it!
+
+For more information on the packs, skip ahead to the strategies section.
```
Comment (ulfarsson) [inline README.rst:226]: "The last argument above `point_placements` is a strategy pack, which we explain in detail in the StrategyPacks section"

File: README.rst [hunk 17]
```diff
@@ -58,17 +69,19 @@ denote the cells in which the points of ``π`` are drawn on a grid. Let
 permutations is defined the same as containment of permutations, except
 including the preservation of the cells.
 
+# TODO: add an example
```
[the remark this answers is not in the document]
Comment (christianbean) [reply README.rst:72]: Would this work? https://github.com/PermutaTriangle/Tilings/tree/readme-image-test 
I can merge it in if we think so.

File: README.rst [hunk 18]
```diff
@@ -158,5 +179,938 @@ using the ``gridded_perms_of_length`` method.
...
+
+StrategyPacks
+=============
+
+We have implemented a large number of structural decomposition strategies that
+we will discuss a bit more in the strategies section that follows. One can use
+any subset of these strategies to search for a combinatorial specification.
+This can be done by creating a ``TileScopePack``.
+
+We have prepared a number of curated pack of strategies that we find to be
```
Comment (ulfarsson) [inline README.rst:413]: packS

File: README.rst (hunk 12 above)
Comment (christianbean) [reply README.rst:145]: Added

File: README.rst (hunk 13 above)
Comment (christianbean) [reply README.rst:158]: fixed

File: README.rst [hunk 19]
```diff
@@ -158,5 +179,938 @@ using the ``gridded_perms_of_length`` method.
...
+       ...     print(strategy(tiling))
+       insert 0 in cell (0, 0)
+       +-+            +-+     +-+
+       |1|         =  | |  +  |1|
+       +-+            +-+     +-+
+       1: Av(120)             1: Av+(120)
+                              Requirement 0:
+                              0: (0, 0)
+
+The same underlying principal corresponds to avoiding or containing any set of
```
Comment (ulfarsson) [inline README.rst:733]: principle

File: README.rst (hunk 15 above)
Comment (christianbean) [reply README.rst:221]: fixed to  "by default this command will ..."

File: README.rst (hunk 18 above)
Comment (christianbean) [reply README.rst:413]: fixed

File: README.rst (hunk 19 above)
Comment (christianbean) [reply README.rst:733]: fixed

File: README.rst [hunk 20]
```diff
@@ -158,5 +179,938 @@ using the ``gridded_perms_of_length`` method.
...
+
+
+Row and column separation
+-------------------------
+
+Every non-empty permutation in ``Av(231)`` can be written in the form αnβ where
+``α``, ``β`` are permutation avoiding ``231``, and all of the values in ``α``
+are below all of the values in ``β``. The tiling representing placing the
+topmost point in ``Av(231)`` contains a crossing size 2 obstruction
+``10: (0, 0), (2, 0)``. This obstruction precisely says that the points cell
```
Comment (ulfarsson) [inline README.rst:876]: point cell

File: README.rst [hunk 21]
```diff
@@ -158,5 +179,938 @@ using the ``gridded_perms_of_length`` method.
...
+
+We instead use the symbol ``*`` to make us aware that this is not counted
+by the Cauchy product, but we must also count the possible interleavings.
+
+
+Obstruction inferral
+--------------------
+
+The presence of requirements alongside the obstructions on a tiling can
+sometimes be use to imply that all of the gridded permutations on a tiling also avoid
```
Comment (ulfarsson) [inline README.rst:989]: useD

File: README.rst (hunk 14 above)
Comment (christianbean) [reply README.rst:181]: fixed

File: README.rst (hunk 16 above)
Comment (christianbean) [reply README.rst:226]: edited to 
"The ``point_placements`` argument above is a strategy pack, which we explain in
more detail in the ``StrategyPacks`` section."

File: README.rst (hunk 21 above)
Comment (christianbean) [reply README.rst:989]: fixed

File: README.rst (hunk 20 above)
Comment (christianbean) [reply README.rst:876]: fixed

File: README.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (jaypantone) [reply README.rst:383]: Very nice

File: README.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (jaypantone) [reply README.rst]: Great

File: README.rst (hunk 17 above)
[the remark this answers is not in the document]
Comment (jaypantone) [reply README.rst:72]: I think the ascii version is excellent.

File: README.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (jaypantone) [reply README.rst]: What you added is good.

File: README.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (christianbean) [reply README.rst:383]: I forgot to add the sentence plugging our discord server again! I'll do that.

[the remark this answers is not in the document]
Comment (christianbean) [reply README.rst:383]: I added this sentence to the end of the section.
"If you have a system of equations you are unable to solve, then please feel
free to send them to our `Discord server <https://discord.gg/ySJD6SV>`__."

File: README.rst [hunk 22]
```diff
@@ -19,8 +19,18 @@ Tilings
...
+can be used to enumerate permutation classes.
+
+If you are primarily interested in enumerating permutation classes, then you
+may wish to skip ahead to the ``TileScope`` section, but note the installation
+will be the same as for ``tilings``.
+
+If you need support, have a suggestion, or just want to be up to date with the
+latest developments please join us on our
+`Discord server <https://discord.gg/ySJD6SV>`__ where we'd be happy to hear
+from you!
```
[the remark this answers is not in the document]
Comment (christianbean) [reply README.rst:34]: Perhaps we should have a section on "Performance" where we can discuss pypy and also the extra commands to give to python or pypy for help with memory issues.

**Comment by jaypantone** [comment]
I haven't had time to address the particular comments yet, but had a random thought. Maybe we should add something near the top like "If you want to be notified via email of major releases, ..."

**Comment by christianbean** [comment]
> I haven't had time to address the particular comments yet, but had a random thought. Maybe we should add something near the top like "If you want to be notified via email of major releases, ..."

We can do that and leave the permuta triangle email?

**Comment by christianbean** [comment]
> Also, side note, sometimes you get a different tree for Av(123) than the one in the readme, so it thinks it's broken. If that happens, just restart the test and pray to the RNG.

I haven't seen this happen yet. However, if it becomes a real issue, we can change it to placing only top or only bottommost rows, then we can be more confident.

File: README.rst [hunk 23]
```diff
@@ -157,6 +202,961 @@ using the ``gridded_perms_of_length`` method.
...
+       \: Av(01)                       \: Av(01)
+       Crossing obstructions:          Crossing obstructions:
+       01: (0, 0), (1, 0)              012: (0, 0), (1, 0), (1, 0)
+       012: (0, 0), (2, 0), (2, 0)     Assumption 0:
+       012: (1, 0), (2, 0), (2, 0)     can count occurrences of
+                                       0: (0, 0)
+
+=========
+
+Finally, I'd like reiterate, if you need support, have a suggestion, or just
```
Comment (ulfarsson) [inline README.rst:1159]: Missing "to". Also, perhaps say "we would like to reiterate..." I think "we" has always been used above.

Comment (jaypantone) [reply README.rst]: Fixed.

File: README.rst (hunk 22 above)
[the remark this answers is not in the document]
Comment (jaypantone) [reply README.rst:37]: Added.

File: README.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (jaypantone) [reply README.rst]: Made small adjustments.

File: README.rst [hunk 24]
```diff
@@ -19,8 +19,21 @@ Tilings
...
+may wish to skip ahead to the ``TileScope`` section, but note the installation
+will be the same as for ``tilings``.
+
+If you need support, have a suggestion, or just want to be up to date with the
+latest developments please join us on our
+`Discord server <https://discord.gg/ySJD6SV>`__ where we'd be happy to hear
+from you! To receive an email notification about major new releases,
+send an email to `permutatriangle@gmail.com <mailto:permutatriangle@gmail.com>`_
+(but please direct all requests for assistance to the
+`Discord server <https://discord.gg/ySJD6SV>`__).
```
Comment (jaypantone) [inline README.rst:36]: Here's where I added the email.

File: README.rst [hunk 25]
```diff
@@ -157,6 +207,1002 @@ using the ``gridded_perms_of_length`` method.
...
+       0: (1, 2)
+
+The ``x`` in the printed above rule is used to denote Cartesian product.
+We do this to signify that there is a size-preserving bijection between the
+gridded permutations on the left-hand side, to the set of 3-tuples coming from
+the Cartesian product on the right-hand side, where the size of a tuple is the
+sum of the sizes of the parts. In particular, it implies that the enumeration
+of the gridded permutations on the left-hand side can be computed by applying the
+Cauchy product to the enumerations of the three sets of gridded permutations on
+the right-hand side.
```
Comment (jaypantone) [inline README.rst:1005]: I added the last sentence.

File: README.rst [hunk 26]
```diff
@@ -157,6 +207,1002 @@ using the ``gridded_perms_of_length`` method.
...
+       >>> from comb_spec_searcher import CombinatorialSpecification
+       >>> reloaded_spec = CombinatorialSpecification.from_dict(json.loads(json_string))
```
Comment (jaypantone) [inline README.rst:470]: New JSON section

File: README.rst [hunk 27]
```diff
@@ -157,6 +207,1002 @@ using the ``gridded_perms_of_length`` method.
...
+.. code::
+
+    PYPY_GC_MAJOR_COLLECT=1.1
+    PYPY_GC_MAX_DELTA=200MB
+    PYPY_GC_INCREMENT_STEP=10GB
+
+tends to improve memory usage at the cost of 30% - 50% extra time.
+
+If memory usage, rather than time usage, is a bottleneck, then the default interpreter
+``CPython`` is preferred.
```
Comment (jaypantone) [inline README.rst:1201]: New performance section

Comment (jaypantone) [reply README.rst:1201]: Oops, I'll fix "mitigating" -> "mitigated"

**Comment by jaypantone** [comment]
> > Also, side note, sometimes you get a different tree for Av(123) than the one in the readme, so it thinks it's broken. If that happens, just restart the test and pray to the RNG.
> 
> I haven't seen this happen yet. However, if the issue keeps coming up, then we can change it to placing only top or only bottommost rows, then we can probably be more confident.

https://travis-ci.org/github/PermutaTriangle/Tilings/jobs/697715681

**Pull request merged**